gio-subclass/src/action.rs

```diff
@@ -0,0 +1,220 @@
+}
+
+unsafe extern "C" fn action_get_state<T: >
+(ptr: *mut GAction) -> *mut glib::GVariant
+where
+    T::ImplType: ActionImpl<T>
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.get_state()
+}
+
+unsafe extern "C" fn action_get_state_hint<T: >
+(ptr: *mut GAction) -> *mut glib::GVariant
+where
+    T::ImplType: ActionImpl<T>
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.get_state_hint()
+}
+
+unsafe extern "C" fn action_get_state_type<T: >
+(ptr: *mut GAction) -> *const glib::GVariantType
+where
+    T::ImplType: ActionImpl<T>
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.get_state_type()
+}
+
+unsafe extern "C" fn action_init<T: ObjectType>(
+    iface: glib_ffi::gpointer,
+    iface_data: glib_ffi::gpointer
+) {
+    callback_guard!();
+    let action_iface = &mut *(iface as *mut ffi::);
+    let iface_type = (*(iface as *const gobject_ffi::GTypeInterface)).g_type;
+    let type_ = (*(iface as *const gobject_ffi::GTypeInterface)).g_instance_type;
+    let klass = &mut *(gobject_ffi::g_type_class_ref(type_) as *mut ClassStruct<T>);
+    let interfaces_static = &mut *(klass.interfaces_static as *mut Vec<_>);
+    interfaces_static.push((iface_type, iface_data))
+    action_iface.activate = Some(action_activate::<T>);
+    action_iface.change_state = Some(action_change_state::<T>);
+    action_iface.get_enabled = Some(action_get_enabled::<T>);
+    action_iface.get_name = Some(action_get_name::<T>);
+    action_iface.get_parameter_type = Some(action_get_parameter_type::<T>);
+    action_iface.get_state = Some(action_get_state::<T>);
+    action_iface.get_state_hint = Some(action_get_state_hint::<T>);
+    action_iface.get_state_type = Some(action_get_state_type::<T>);
+}
+
+unsafe extern "C" fn action_get_type<T: ObjectType>(
+    type_: glib_ffi::GType
+) -> glib::Type {
```

Comment:
This function is not needed for interfaces, what you basically have here is a trampoline calling a `get_type` function on the interface... but there is none on `GAction`